### PR TITLE
Fix gradient and var name mismatch when num_groups > 0

### PR DESCRIPTION
**Pull request by Infi-zc**
## Checklist before submitting

- [x ] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes [Mismatch of grads and vars in group allreduce](https://github.com/horovod/horovod/issues/3458)

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: horovod/tensorflow/__init__.py [hunk 1]
```diff
@@ -495,6 +496,8 @@ def compute_gradients(self, *args, **kwargs):
             """
             gradients = self._optimizer.compute_gradients(*args, **kwargs)
             grads, vars = zip(*gradients)
+            if self._group is not None and not isinstance(self._groups, list) and self._group > 0:
```
Comment (maxhgerlach) [inline horovod/tensorflow/__init__.py:499]: Could you motivate the logic in this condition a bit?

`self._group` is probably a typo for `self._groups`?

Is this fix not necessary for the case where `self._groups` is a list?

**Review by maxhgerlach — CHANGES_REQUESTED** [review]
Thanks for giving this a go! 

It would be good to add a unit test to `tests/parallel/test_tensorflow.py` that fails without the fix.

Note that you would need to sign the [Developer Certificate of Origin](https://probot.github.io/apps/dco/) by [signing off your commits](https://github.com/horovod/horovod/pull/3436/checks?check_run_id=5376922918).

File: horovod/tensorflow/__init__.py (hunk 1 above)
Comment (Infi-zc) [reply horovod/tensorflow/__init__.py]: Sorry for the typo here, I have corrected it. 
Yes, I think this fix is unnecessary for the case where self._groups is a list. In this case (self._groups: list) the code handles both var name and grad, which will not lead to the error discribed in the issue. 
Next I will try to add a unit test to test_tensorflow.py to reproduce the problem.

**Comment by EnricoMi** [comment]
@maxhgerlach @Infi-zc what is the status of this PR?

**Comment by maxhgerlach** [comment]
I haven't quite understood why the problem does not appear in case the `groups` are specified as an explicit list (there is some related discussion in the PR that introduced the feature: https://github.com/horovod/horovod/pull/2523#issuecomment-771391796). A test that showcases the problem could be helpful here.

**Pull request closed without merging**